src/text/line.rs

```diff
@@ -586,31 +553,55 @@ impl WidgetRef for Line<'_> {
     fn render_ref(&self, area: Rect, buf: &mut Buffer) {
         let area = area.intersection(buf.area);
         buf.set_style(area, self.style);
-        let width = self.width() as u16;
-        let mut x = area.left();
-        let line = if width > area.width {
-            self.truncated(area.width)
+        let line_width = self.width() as u16;
+        if line_width <= area.width {
+            let offset = match self.alignment {
+                Some(Alignment::Center) => (area.width.saturating_sub(line_width)) / 2,
+                Some(Alignment::Right) => area.width.saturating_sub(line_width),
+                Some(Alignment::Left) | None => 0,
+            };
+            render_spans(&self.spans, area, buf, offset, 0);
         } else {
             let offset = match self.alignment {
-                Some(Alignment::Center) => (area.width.saturating_sub(width)) / 2,
-                Some(Alignment::Right) => area.width.saturating_sub(width),
+                Some(Alignment::Center) => (line_width.saturating_sub(area.width)) / 2,
+                Some(Alignment::Right) => line_width.saturating_sub(area.width),
                 Some(Alignment::Left) | None => 0,
             };
-            x = x.saturating_add(offset);
-            self.to_owned()
+            render_spans(&self.spans, area, buf, 0, offset);
         };
-        for span in &line.spans {
-            let span_width = span.width() as u16;
-            let span_area = Rect {
-                x,
-                width: span_width.min(area.right().saturating_sub(x)),
-                ..area
-            };
-            span.render(span_area, buf);
-            x = x.saturating_add(span_width);
-            if x >= area.right() {
-                break;
-            }
+    }
+}
+
+fn render_spans(
+    spans: &[Span],
+    area: Rect,
+    buf: &mut Buffer,
+    mut x_offset: u16,
+    mut span_offset: u16,
```

Comment:
Is `x_offset` the byte offset and `span_offset` the offset in chars?

Reply:
The other way around - likely a better name would be `buffer_x_offset` perhaps?

Reply:
Rect needs an easy way to increase the x value and decrease the width the same amount, but I'm not sure of a good name for it.

Having that method would make for loops that have to be area constrained, but that write across horizontally easier to write, and the x_offset parameter would then go away and be already part of the area passed into the function.

Something like 
```rust
area.indent_x(span_width);
```

Reply:
addressed.